File: python/examples/example-socket-publisher.py
# SPDX-License-Identifier: BSL-1.0

import time
import traeger


def broadcast(scheduler: traeger.Scheduler, publisher: traeger.Publisher, counter: int):
    print(f"Broadcasting heart-beat {counter}")
    publisher.publish(scheduler, "heart-beat", counter)
    if counter < 10:
        heart_beat(scheduler, publisher, counter + 1)


def heart_beat(
    scheduler: traeger.Scheduler, publisher: traeger.Publisher, counter: int
):
    scheduler.schedule_delayed(1, lambda: broadcast(scheduler, publisher, counter))


def main():
    context: traeger.Context = traeger.Context()
    address: str = "tcp://*:5556"
    format: traeger.Format = traeger.Format("json")
    publisher: traeger.Publisher = context.publisher(address, format)

    scheduler: traeger.Scheduler = traeger.Scheduler(threads_count=8)
    print(f"Publishing heart-beat events on address: {address}")
    heart_beat(scheduler, publisher, 0)

    while scheduler.count() != 0:
        time.sleep(0.010)


if __name__ == "__main__":
    main()

File: rust/examples/example-socket-publisher.rs
// SPDX-License-Identifier: BSL-1.0

use std::{thread::sleep, time::Duration};

use traeger::{Context, Format, Int, Publisher, Scheduler, Threads, Variant};

fn heart_beat(scheduler: &Scheduler, publisher: &Publisher, counter: Int) {
    let scheduler_clone = scheduler.clone();
    let publisher_clone = publisher.clone();
    scheduler.schedule_delayed(Duration::from_secs(1), move || {
        println!("Broadcasting heart-beat: {counter}");
        publisher_clone.publish(&scheduler_clone, "heart-beat", Variant::Int(counter));
        if counter < 10 {
            heart_beat(&scheduler_clone, &publisher_clone, counter + 1);
        }
    });
}

fn main() {
    let context = Context::new();
    let address = "tcp://*:5556";
    let format = Format::by_name("json").unwrap();

    match context.publisher(address, &format) {
        Err(error) => {
            println!("Socket bind error: {error}");
        }
        Ok(publisher) => {
            let scheduler = Scheduler::new(Threads { count: 8 });
            println!("Publishing heart-beat events on address: {address}");
            heart_beat(&scheduler, &publisher, 0);

            while scheduler.count() != 0 {
                sleep(std::time::Duration::from_millis(10));
            }
        }
    }
}
